lib/src/path.cc

```diff
@@ -0,0 +1,215 @@
+        return !remainder()._path.isEmpty();
+    }
+
+    shared_string path::last() const {
+        path p = *this;
+        while (p.has_remainder()) {
+            p = p.remainder();
+        }
+        return p.first();
+    }
+
+    path path::prepend(path prefix) {
+        path_builder builder;
+        builder.append_path(prefix);
+        builder.append_path(*this);
+        return builder.result();
+    }
+
+    int path::length() const {
+        int length = 1;
+        path p = *this;
+        while (p.has_remainder()) {
+            length++;
+            p = p.remainder();
+        }
+        return length;
+    }
+
+    path path::sub_path(int remove_from_front) {
+        int count = remove_from_front;
+        path p = *this;
+        while (p.has_remainder() && count > 0) {
+            count--;
+            p = p.remainder();
+        }
+        return p;
+    }
+
+    path path::sub_path(int first_index, int last_index) {
+        if (last_index < first_index) {
+            throw config_exception("Bad call to sub_path: invalid range");
+        }
+
+        path from = sub_path(first_index);
+        path_builder builder;
+        int count = last_index - first_index;
+        while (count > 0) {
+            count--;
+            builder.append_key(*from.first());
+            from = from.remainder();
+            if (from.has_remainder()) {
+                throw config_exception("sub_path last_index out of range");
+            }
+        }
+        return builder.result();
+    }
+
+    bool path::starts_with(path other) const {
+        path my_remainder = *this;
+        path other_remainder = other;
+        if (other_remainder.length() <= my_remainder.length()) {
+            while (other_remainder.first()) {
+                if (*other_remainder.first() != *my_remainder.first()) {
+                    return false;
+                }
+                my_remainder = my_remainder.remainder();
+                other_remainder = other_remainder.remainder();
+            }
+            return true;
+        }
+        return false;
+    }
+
+    path path::parent() const {
+        if (!has_remainder()) {
+            return path { };
+        }
+
+        path_builder builder;
+        path p = *this;
+        while (p.has_remainder()) {
+            builder.append_key(*p.first());
+            p = p.remainder();
+        }
+        return builder.result();
+    }
+
+    bool path::operator==(path const& other) const {
+        if ((first() == nullptr) ^ (other.first() == nullptr)) {
+            return false;
+        }
+        if (first() == nullptr && other.first() == nullptr) {
+            return true;
+        }
+        return *first() == *other.first() &&
+               remainder() == other.remainder();
+    }
+
+    bool path::operator!=(path const& other) const {
+        return !(*this == other);
+    }
+
+    bool path::has_funky_chars(string const& s) {
+        if (s.length() == 0) {
```

Comment:
To me it reads more clearly if this is `if (s.empty()) {` (There might be a performance gain as well, but that's a different conversation.)

If others agree, should the style guide add a preference for using the `empty` method of standard containers over size checks of 0?

Separate question: is this needed here? I.e. will the logic below do the Right Thing with an empty string?